### PR TITLE
Preserve reducer order in schema conversion

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

This was found while working on #1965. The tests will come in that other PR, but it was suggested that we should land the correctness fix itself sooner.

# API and ABI breaking changes

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

File: crates/schema/src/def.rs [hunk 1]
```diff
@@ -84,7 +85,9 @@ pub struct ModuleDef {
     tables: IdentifierMap<TableDef>,
 
     /// The reducers of the module definition.
-    reducers: IdentifierMap<ReducerDef>,
+    /// Note: this is using IndexMap because reducer order is important
```
Comment (kazimuth) [inline crates/schema/src/def.rs:88]: Can we get a corresponding comment on `RawModuleDefv9` as well

**Review by kazimuth — COMMENTED** [review]
This looks good, I think will need a corresponding `Private` PR.

This *is* inconsistent with how IDs work for other entity types.
Two alternatives would be:
- Provide a `_reducer_id_from_name` API similar to indexes/tables
- Require the user to sort their reducers when computing IDs

This is definitely the easiest patch though.

**Comment by RReverser** [comment]
> * Require the user to sort their reducers when computing IDs

Do you mean with validation check? That would also be fine by me as an alternative.

File: crates/schema/src/def.rs (hunk 1 above)
Comment (RReverser) [reply crates/schema/src/def.rs:88]: Where do you mean? I didn't change its type.

Comment (kazimuth) [reply crates/schema/src/def.rs:88]: For most of the vectors in that struct, order does not matter, the contained data is sorted. However, this is no longer true for the `reducers` vector, whose order is now semantically important.

**Comment by kazimuth** [comment]
Yeah, I think that's less confusing overall. I can do another branch that adds that check in `schema`.

**Comment by RReverser** [comment]
If you think that's better, sure, go for it.

Personally I don't care too much. I do think that preserving whatever order happens to be, is easier for module binding libraries, particularly because a validation error would be visible a bit late in the pipeline (user would get an error only when an example module with mismatched order is compiled & published), but given that we only have 2 such libraries and not creating new one every month or so, it shouldn't be a big deal.

File: crates/schema/src/def.rs (hunk 1 above)
Comment (kazimuth) [reply crates/schema/src/def.rs:88]: Actually, this sort of doesn't matter, since the module itself never uses the Reducer ID.

**Comment by kazimuth** [comment]
Ah, no, you're right. I was confused -- it really doesn't matter inside the module, the module never uses reducer IDs. It's just an interface somewhere else. So, I'm happy with this.

**Review by kazimuth — APPROVED** [review]
LGTM with tests passing

**Comment by RReverser** [comment]
That's... some new test failure.

> This looks good, I think will need a corresponding `Private` PR.

Also, does this still stand / do I need to change private too?

**Comment by kazimuth** [comment]
It looks like both test failures here were caused by unrelated heisenbugs on master, we don't need a Private fix. This one might also go away on rerun.

**Pull request merged**